### ISSUE TITLE
Amazfit Bip does not reconnect (only cycling BT works!)

### ISSUE THREAD

**Issue by TimH54** [Freeyourgadget/Gadgetbridge#1142]
#### Before opening an issue please confirm the following:
- [x] I have read the [wiki](https://github.com/Freeyourgadget/Gadgetbridge/wiki), and I didn't find a solution to my problem / an answer to my question.
- [x] I have searched the [issues](https://github.com/Freeyourgadget/Gadgetbridge/issues), and I didn't find a solution to my problem / an answer to my question.

#### Your issue is:

My AmazFit Bip works nicely with Gadgetbridge, except for one serious problem: If the connection to my phone is lost, e.g. because the phone is shortly out of range, GadgetBridge is not able to reconnect successfully to the watch (the message "Connecting" remains indefinitely). The problem happens reproducably both for the automatic reconnect and when initializing the reconnect manually from the main screen. What does work - every time - is cycling Bluetooth. It takes up to 30 seconds, but the connection works every time and is then also stable.

I noticed that the reconnect (without cycling BT) does seem to work to some extent:
* the watch reports an established connection (at least for a minute or so)
* the phone's "smart lock" functionality unlocks the phone, i.e., the watch is connected at phone level, even if Gadgetbridge does not fully establish a connection
* Gadgetbridge reports the battery level on the main screen, but nothing more (notifications do not come through either)

I have tried several firmware versions for the watch: 0.1.1.14, 0.1.1.45 and the brandnew 1.0.2.0. They all behave the same.

The logfile has me first successfully connecting the watch after switching on BT. Then I force a disconnect by moving the phone away from the watch. The following reconnect does not come through, as described above. Finally, I cycle BT and the watch connects successfully.

#### Your wearable device is:

Amazfit Bip HW version 0.11.2.4, various firmwares (see above).

#### Your android version is:

Android 8.0.0 (Oneplus 3T OxygenOS)

#### Your Gadgetbridge version is:

0.27.1 and 0.27.0
[gadgetbridge.log](https://github.com/Freeyourgadget/Gadgetbridge/files/2173844/gadgetbridge.log)

**Comment by TimH54** [comment]
Hi again,

my earlier statement that the connection - once established - is stable, unfortunately turns out not to be true. A few times a day the connection between the Bip and Gadgetbridge drops, even though the phone was always closeby. What is really bad about this is that the watch still reports the connection to be established, so a quick glance at the watch does not tell you that the connection was lost and thus notifications will not get through.

PS: I should have said so before: Gadgetbridge is a fantastic project! I have been using it for my Pebble Time Steel, and am really happy that all my private data remains locally on my phone. Thank you, and keep up the good work!

**Comment by IzzySoft** [comment]
> A few times a day the connection between the Bip and Gadgetbridge drops

I see something similar with my Pebble Time Steel as well. Difference is that the PTS shows the connection has dropped – and a little later it is reestablished automatically. Distance between my Android device and the PTS is rarely more than a meter (PTS at my arm, phone at my belt or in my hand).

**Comment by IzzySoft** [comment]
Good points, @TimH54 – though one can use the PTS in BLE mode as well. I've tried that once, but it wasn't really satisfying (AFAIR it was too frequent connection losses sometimes requiring manual reconnect, also resulting in higher battery consumption on the PTS which contradicts the reasoning for BLE).

OK, I better shut up here then: if it's a different case, my quack would rather make the issue harder to follow :wink:

**Comment by danielegobbetti** [comment]
For me the ble mode (pebble time, omnirom android 8.1) was completely unreliable until 0.26, then when developing the client only mode (that works with p2 **only**) we made some changes that made the connection rock solid. That is to say that if you had troubles in the past you could give it another spin. 

Back on topic, Regarding the bip issues, can you please share whether the automatic fetch on unlock was activated? Could you try toggling it and see if something changes?

Thanks a lot for the kind words!

**Comment by IzzySoft** [comment]
@danielegobbetti my last try was with 0.26.5 and for about 8 hours, with ~3 connection losses. Connection was gone, reconnect took ages if it worked at all. In the end it did NOT reconnect, but the Pebble still showed it were connected (so it could indeed be related); GB hung on "waiting for reconnect". If you know anything I should try please let me know and I'll give it another run. I didn't delete the BLE entry, just unset to prefer it.

**Comment by IzzySoft** [comment]
@TimH54 

> If I understood you correctly, the issues appear when you switch the connection to the PTS to BLE.

Exactly. And only then. With normal BT the connection seems to be lost every now and then as well (which is a different issue), but I rarely notice as a) if I notice, I see the PTS having noticed it and b) it immediately reconnects.

TL;DR: My PTS shows the same symptoms when in BLE mode.

**Comment by TimH54** [comment]
> It was switched off so far. I have now switched it on and will report back if it helps after trying it for a day or so.

Ok, no need to test any further: This does not make any difference. Once the connection is lost, it cannot be reestablished by Gadgetbridge (while the watch and the phone report that the watch is still connected). Only cycling bluetooth solves the issue.

Please let me know if I can test anything else or provide any further information!

**Comment by Kiolali** [comment]
I have the same problem! (OnePlus 3T, Android 8.0.0)

**Comment by aldem** [comment]
I have a similar issue with AmazFit Bip (fw 1.1.2.05, hw 0.11.1.4, on Note 8 Android 8.0) - once phone was out of range for a while, connection to GB is lost, though watch indicates that it is reestablished.

Then GB shows message "Fetching activity data" and nothing happens (with or without "Auto fetch" option).

Long press to disconnect and short to connect again works (no BT cycling is needed).

**Comment by Pepe404** [comment]
Same problem here, LG G2, LineageOS 15.1. After the connection is lost, it will never reconnect automatically. Makes the Gadgetbridge completely unreliable, sorry.

**Comment by IzzySoft** [comment]
@Pepe404 see the comment just before yours. If 0.29 doesn't fix it for you, you're also welcome to send in logs :wink:

**Comment by Pepe404** [comment]
But there is no 0.29 yet, right?

**Comment by ashimokawa** [comment]
I know why I ignore this issue tracker.

**Comment by TimH54** [comment]
Hey guys, it's great to see some (potential) progress on this! I will wait for the new release to appear on F-Droid and then test it as soon as I again have access to my Bip (which unfortunately is only as of 18th August). I'll be sure to report back here. In the meantime, let's best stay on topic. ;)

**Comment by IzzySoft** [comment]
@Pepe404 So? How do you expect it to be fixed without coding? Carsten wrote he has worked on it and it's hopefully fixed with the upcoming new version. So I'd suggest you wait for that version, install that, and verify. If it's solved for you, enjoy – otherwise please send in logs. That's what I was saying. Sorry if I was unclear :innocent: 

Hint: v0.29 is not tagged yet, so it will take at least a couple of days before it can popup at F-Droid.

**Comment by danielegobbetti** [comment]
@Pepe404 don't be rude and stay on topic.

**Comment by ashimokawa** [comment]
@Pepe404

About #980. Just recently there were firmware changes on the BIP which broke  weather (latest fw). This is also fixed in master and will be in the next release.

You could at least look at the commit log messages that happened after your current release before nagging about fixed bugs.

**Comment by Pepe404** [comment]
0.29, reconnect works, at least most of time. It happened once for me, that it didn't reconnect, maybe I haven't waited long enough, I reconnected manually, without restarting BT.

**Comment by cpfeiffer** [comment]
@Pepe404 Thanks for testing, so I'll close this issue. If it happens again, I'm pretty sure it will have a different cause. In that case, please open a new issue and attach a log.

**Comment by Pepe404** [comment]
0.29.1 - this issue still persists, maybe 30% of the time, after going out of BT range, it will not reconnect. No need to cycle BT to reconnect, only needs to tap to connect in GB (when it shows "not connected"), or restart GB (if it still shows "connected", but BT icon on the watch is red, and there is no conennection). LineageOS 15.1, Bip firmware 1.1.2.05. Bip may be paired or unpaired in system BT settings, I can't see any relation, problem happens in both cases.

**Comment by TimH54** [comment]
I wasn't aware one could still comment after the issues is closed, otherwise I would also have reported back.

For me, things seem to work rather stably. I only sometimes found GadgetBridge stuck on "fetching activity data" (I had automatic fetching on unlock every 60 minutes enabled). I have now set a much larger time window for automatic fetching and so far, things seem to be ok.

@Pepe404 I guess if the problem persists, you need to reproduce the problem and provide a log file ...

**Comment by TimH54** [comment]
@Pepe404 I have seen connection hanging a few times now as well. Will try to catch it with a log file. Once I do, I will open a new issue (if you have not), as I think the devs do not watch closed issues any more.

**Comment by Pepe404** [comment]
Yes, this problem really exists, I'm tired of supervising this watch, "Is it still connected? Why no new notifications for few hours?", very unreliable, I reinstalled Mi Fit, zero problems with silent connection hangs, and reconnection is always very quick. Will check the future release of GB, if the problem is fixed or not.

**Comment by Pepe404** [comment]
I'm afraid, that this reverse engineering will never be 100% successful. In the meantime, Mi Fit was significantly improved, no more delayed/missed/repeated notifications, instant and reliable reconnection, no significant battery drain. Ok, my private data is uploaded somewhere to Chinese Xiaomi servers, in fact all my private data is already uploaded to American or Jewish Google servers, so actually I have nothing really private on my phone, no worries.

**Comment by cutec-chris** [comment]
I have seen this issue also 2 times a day now.
